# Assignment 9 - Basic Machine Learning Example
### Let's Play with a Decision Tree 
Let's try and predict housing prices in Iowa using a Test & Train dataset provided by Kaggle

Source & images for the notebook: **kaggle.com/learn/intro-to-machine-learning.**

#### Your Challenge - predicting sales price
We'll explore a dataset for houses in Iowa that has a variety of different attributes (called "features" in the ML world). 

Your goal is to try and figure what a range for a sales price might look like.

# Grade: 100%

### Assignment
1. Explore the Iowa housing data (EDA) using Pandas methods
2. Pick what features you think might better predict price as measured by MEA
3. Split the data (test, train, validate)
4. Predict your values & train your model
5. Validate your predictions
6. Run a random forest to compare decision tree

### Decision Trees
There are a variety of model to use with machine learning. First up, we'll use a tool called decision tree which has "leaves" and "nodes". Basically, it splits data at key parts (nodes) based on an attribute (leaf) and then goes further and further down the branch. 

A good visualization here:
<img src="https://res.cloudinary.com/dyd911kmh/image/upload/f_auto,q_auto:best/v1545934190/1_r5ikdb.png">
Source: https://www.datacamp.com/community/tutorials/decision-tree-classification-python

However, at some point, the leaves can be so small that the predictions match the actual values. This is called "overfitting" because the predictions overfit for new data.

On the other side, we can divide on just a few leaves and what happens is that the model fails to capture important distinctions or patterns in the data. It will perform poorly on training data and is called "underfitting".

<img src="https://i.imgur.com/2q85n9s.png">

Knowing how to fine tune all this is where the art of machine learning comes in using your data and what you have at hand (called "features" or "feature engineering"). The science of evaluating your model comes in with using statistical analysis. There's always a tension here.

Image and Content Source: https://www.kaggle.com/dansbecker/underfitting-and-overfitting

In this example, we'll use a decision tree to test out predicting price for house prices in Iowa (the dataset we'll be using). You can think of this example for how we might split housing data in the following way: 

<img src="https://i.imgur.com/R3ywQsR.png">
Source: https://www.kaggle.com/learn/intro-to-machine-learning

In [1]:
# Path of the file to read of Iowa housing data set
iowa_file_path = '/users/steviodong/Desktop/msba_class/msba_655/msba_python/msba-python/09_numpy/train.csv'

In [2]:
# Importing our standard libraries
import pandas as pd
import numpy as np

# Loading our machine learning library from Scikit Learn
from sklearn.tree import DecisionTreeRegressor

# Loading mean absolute error
from sklearn.metrics import mean_absolute_error

# Loading test/train/split library
from sklearn.model_selection import train_test_split

# Setting our columns to max so we can see everything 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# read & store our data in a DataFrame called home_data
home_data = pd.read_csv(iowa_file_path)
home_data.head(5)
home_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.549315   567.240411  1057.429452  1162.626712   346.992466   
std     161.319273   441.866955   438.705324   386.587738   436.528436   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   223.000000   795.750000   882.000000     0.000000   
50%       0.000000   477.500000   991.500000  1087.000000     0.000000   
75%       0.000000   808.000000  1298.250000  1391.250000   728.000000   
max    1474.000000  2336.000000  6110.000000  4692.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000   
mean       5.844521  1515.463699      0.425342      0.057534     1.565068   
std       48.623081   525.480383      0.518911      0.238753     0.550916   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1129.500000      0.000000      0.000000     1.000000   
50%        0.000000  1464.000000      0.000000      0.000000     2.000000   
75%        0.000000  1776.750000      1.000000      0.000000     2.000000   
max      572.000000  5642.000000      3.000000      2.000000     3.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000   
mean      0.382877      2.866438      1.046575      6.517808     0.613014   
std       0.502885      0.815778      0.220338      1.625393     0.644666   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     14.000000     3.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean   1978.506164     1.767123   472.980137    94.244521    46.660274   
std      24.689725     0.747315   213.804841   125.338794    66.256028   
min    1900.000000     0.000000     0.000000     0.000000     0.000000   
25%    1961.000000     1.000000   334.500000     0.000000     0.000000   
50%    1980.000000     2.000000   480.000000     0.000000    25.000000   
75%    2002.000000     2.000000   576.000000   168.000000    68.000000   
max    2

### Exploring our dataset (EDA)

In [3]:
# Let's explore what our columns include (ideally, we'd have a data dictionary here)

home_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [4]:
# Let's take a quick peek to see if we have any missing values in our columns
# This is critical for picking our features to make sure we don't include holes

home_data.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinSF1          0
BsmtFinType2       38
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
1stFlrSF            0
2ndFlrSF            0
LowQualFin

### Selecting our Prediction Target
We are interested in predicting a sales price - Column name 'SalePrice'

In [5]:
# Let's set our prediction variable or target variable "y" for SalePrice
# Note this is in lower case y - standard practice for ML

y = y = home_data.SalePrice#insert dataframe column you want to predict

### Selecting our Features
These are the parts of our dataset we feel could be useful for predicting price. We'll go with some standard guesses here from how real estate is valued

In [6]:
# Picking standard features for our "X" variable

# Let's start with a custom list of columns that could help us predict sale price based on what's in our dataset
home_data_features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']#Insert your features here from which columns you think will predict price

# Passing these features into our X variable (notice the capital X)
X = home_data[home_data_features]

# Some quick summary statistics
X.describe()

LotArea    YearBuilt     1stFlrSF     2ndFlrSF     FullBath  \
count    1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean    10516.828082  1971.267808  1162.626712   346.992466     1.565068   
std      9981.264932    30.202904   386.587738   436.528436     0.550916   
min      1300.000000  1872.000000   334.000000     0.000000     0.000000   
25%      7553.500000  1954.000000   882.000000     0.000000     1.000000   
50%      9478.500000  1973.000000  1087.000000     0.000000     2.000000   
75%     11601.500000  2000.000000  1391.250000   728.000000     2.000000   
max    215245.000000  2010.000000  4692.000000  2065.000000     3.000000   

       BedroomAbvGr  TotRmsAbvGrd  
count   1460.000000   1460.000000  
mean       2.866438      6.517808  
std        0.815778      1.625393  
min        0.000000      2.000000  
25%        2.000000      5.000000  
50%        3.000000      6.000000  
75%        3.000000      7.000000  
max        8.000000     14.000000

### Picking a Model and Fitting It
At this point, we'll use a decision tree regressor from Scikit Learn machine learning library we loaded up above. Next, we'll fit a our data to this model using the variables we defined for features (X) and our prediction target (y)

In [7]:
# Calling in our preferred method of a Decision Tree
your_variable = DecisionTreeRegressor(random_state = 0) # =0 ensures same results in each run

# Passing in our feature and predicted variables as we defined above
your_variable.fit(X,y)

DecisionTreeRegressor(random_state=0)

In [8]:
# Let's make some predictions using our features defined as "X"
# Returns a Numpy array

your_variable.predict(X)

array([208500., 181500., 223500., ..., 266500., 142125., 147500.])

### Model Validation
How accurate is our model for predicting future prices? There are ton of ways to assess this but in this quick example we'll use Mean Absolute Error (MAE). Our prediction error is basically: error = actual - prediction. For example, if a house costs 500,000 and we predicted 400,00 our error is 100,000. With MAE we take the absolute value which converts errors to positive values.

In [9]:
# Setting a variable to hold our predicted values
predicted_home_prices = your_variable.predict(X)

# Using sklearn method on our target and predicted values
print(mean_absolute_error(y, predicted_home_prices))

#Print your prediction:


62.35433789954339


#### WARNING! 
The above example has a critical flaw - we used the same data mixed together to build the model and evaluate it. This gives us false confidence for accuracy with our predictions because the model has seen this data before. But when we apply it to a dataset we've never seen, we can easily get highly inaccurate data.

The power of ML is to train a model on a dataset, tune it, and then let it make predictions on new datasets.

To do this, we need to split out some of our data as "validation data".

## Splitting Data for Better Accuracy

In [10]:
# First, we've split our data into training and validation data - for BOTH features and target (X, y)
# Note: the split is based on a random number generator. Again, we pass in a numeric value to
# the random_state argument which guarantees we get the same split every time we
# run this script - key for consistency

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

# Define model
your_model = DecisionTreeRegressor(random_state=0)

# Fit model
your_model.fit(train_X, train_y)

# get predicted prices on validation data
your_predictions = your_model.predict(val_X)
print(mean_absolute_error(val_y, your_predictions))

32410.824657534245


#### What This Means
So, when we use our in-sample data, your likely had an MAE of low dollars. Out of sample, it went to a much bigger number - Hint: it should be a huuuge difference! 

Note: our average home value in our dataset is 180921. 

## Tuning our Model for Over/Under Fitting
Here, we'll look at test/train/validate our dataset and then compare our MAE scores to see how many leaves we should include to avoid over or under fitting our model. Note: source for the code and inspiration here: https://www.kaggle.com/dansbecker/underfitting-and-overfitting

<img src="https://i.imgur.com/2q85n9s.png">


Image and Content Source: https://www.kaggle.com/dansbecker/underfitting-and-overfitting


In [15]:
# Let's write a custom function for MAE (source: Kaggle.com) 

def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y): #passing in max_leaf_nodes
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0) #Setting a variable to hold leaves
    model.fit(train_X, train_y) #Same as above for fitting to our training data
    preds_val = model.predict(val_X) #Predict on our validation data
    mae = mean_absolute_error(val_y, preds_val) #Passing in our validation and predicted values for sales
    return(mae)


In [12]:
# Setting up various leaves to test for under/over fitting

for max_leaf_nodes in [5, 50, 125, 250, 500, 5000]: # Iterating through a list of various leaf counts
    iowa_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y) #passing our work from above for each leaf
    print("Max leaf nodes: %d \t\t MAE: %d" %(max_leaf_nodes, iowa_mae))

Max leaf nodes: 5 		 MAE: 35190
Max leaf nodes: 50 		 MAE: 27825
Max leaf nodes: 125 		 MAE: 29017
Max leaf nodes: 250 		 MAE: 31738
Max leaf nodes: 500 		 MAE: 32662
Max leaf nodes: 5000 		 MAE: 33382


#### What This Means
Here we see that our MAE goes to 27825 at 50 leaves and get us better accuracy than our original "out of sample" calculation of 32,110. However at 250 leaves it heads the other way and overfits.

## Bonus - Random Forest
As we saw with a decision tree, there are some obvious tradeoffs. Lot's of leaves can have you overfitting the model because the prediction is coming just a few houses at the leaf. But with just a few leaves, it predicts poorly because it's not accounting for the unique characteristics in the data.

To get around this, we'll use the random forest method, which includes many trees many predictions and then averages everything together. Source: Wikipedia
<img src="https://upload.wikimedia.org/wikipedia/commons/7/76/Random_forest_diagram_complete.png">

#### Our Dataset Splits
Quick refresher on our variables that we split up above
* train_X - our training data comprised of features from our housing dataset: 'LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd'

* val_X - our validation dataset that we use to validate the accuracy of our predictions based on the feature set

* train_y - our target prediction (sales price) that we will train our data against

* val_y - our sales price we want to validate against


In [17]:
# Importing the Random Forest model from Scikit Learn
from sklearn.ensemble import RandomForestRegressor

# Setting a variable to hold our data and setting random state to the same number for consistency
forest_model = RandomForestRegressor(random_state=0)

# Fitting our model on our test and train datasets as defined above
forest_model.fit(train_X, train_y)

# Predicting on our validation dataset
iowa_preds = forest_model.predict(val_X)
print("Our MAE is now:")
print("$",mean_absolute_error(val_y, iowa_preds))

Our MAE is now:
$ 23093.063676581867


## Reflection
In a few sentences what did you learn with this notebook and the dataset? How did your feature selection impact your MAE? What additional datasets do you wish you had to better predict?

Finally, what was the hardest part of this assignment and what did you find most interesting?

In [14]:
# I feel like we are just start to getting in on the fundamental part of data anlytic - building a model. Moreover, the ability to use Python Panda to analysis is far more superior than what we have covered so far. This is a great step and necessary knowledge to learn for our final project. The selection directly affect the result of the MAE, the more the factors there are the more accurate the MAE is for the predict model. I think build upon the model, and further develop the model is the hardest part, but at the same time the result make the whole process most interesting. Moreover, be able to learn from the result make it more fun. 
